R/paginate_listing.R

```diff
@@ -49,20 +49,27 @@ paginate_listing <- function(lsting,
                              tf_wrap = !is.null(max_width),
                              rep_cols = NULL,
                              max_width = NULL,
+                             fontspec = font_spec(font_family, font_size, lineheight),
                              verbose = FALSE,
                              print_pages = TRUE) {
-  checkmate::assert_multi_class(lsting, c("listing_df", "list"))
-  checkmate::assert_numeric(colwidths, lower = 0, len = length(listing_dispcols(lsting)), null.ok = TRUE)
+  ## this is wrong!!! but thats a different PR
+  ## needed to support passing a list in after split_into_pages_by_var
+  ## no guarantee colwidths are right for all chunks
+  samp_lsting <- lsting
+  if (!is(lsting, "listing_df"))
+    samp_lsting <- lsting[[1]]
+  if (is.null(colwidths)) {
+    colwidths <- propose_column_widths(samp_lsting, fontspec = fontspec)
+  }
```

Comment:
I do not think this is needed as it is checked in paginate_to_mpfs for each element

Reply:
When paginating the colwidths need to be the same for all pages, doing propose_column_widths separately for each element in the list doesn't have this constraint, so it isn't sufficient/correct.  This isn't either, as my comment notes, but its closer.

Reply:
That isn't true:

```
lyt <- basic_table() %>%
    split_cols_by("ARM") %>%
    split_rows_by("RACE", split_fun = keep_split_levels(c("WHITE", "BLACK OR AFRICAN AMERICAN"))) %>%
    analyze("AGE")

tbl <- build_table(lyt, ex_adsl)

tbl1 <- tbl[1:2,]

tbl2 <- tbl[3:4,]

lst <- list(tbl1, tbl2)

pagres <- paginate_to_mpfs(lst, cpp = 44, verbose = TRUE)

lapply(pagres, function(x) cat(toString(x)))
```

gives us

```
         A: Drug X   B: Placebo
———————————————————————————————
WHITE                          
  Mean     35.85       32.88   
         C: Combination
———————————————————————
WHITE                  
  Mean       31.95     
                            A: Drug X
—————————————————————————————————————
BLACK OR AFRICAN AMERICAN            
  Mean                        34.06  
                            B: Placebo
——————————————————————————————————————
BLACK OR AFRICAN AMERICAN             
  Mean                        34.93   
                            C: Combination
——————————————————————————————————————————
BLACK OR AFRICAN AMERICAN                 
  Mean                          34.56     
```

So even though these are parts of the same table , it isn't using the same colwidths for both because `propose_column_widths` is empirical and only based on the data it can see. The `WHITE` table only needs to paginate twice  horizontally, while the `BLACK OR AFRICAN AMERICAN` part needs to paginate 3 times. Thus the columns won't line up across the pages.

The code we're commenting on enforces the same colwidths (though, again, not the fully correct ones because it only looks at the first list element, which may not have the widest data in any particular column.

Reply:
That is expected because the input has different tables, and paginate_to_mpfs should not be too clever. If you enforce the use of only one columnwidth, you would have a lot of empty spaces that are not very useful considering they are on different pages. Also, we would lose functionality if we enforced one col width, because it is used now to paginate different elements from a list that has different columnwidths. Hence, I would not assume anything from the list in input. If you want the same value, you can always insert the column_width externally and the values would be fixed for all elements of the list

Reply:
Anyway, I suggest moving this discussion to another issue as this does not concern truetype pagination and is related to another feature (pagination of lists of tables), which is an established and used feature. We can arrange a meeting to discuss this feature in more detail :)

Reply:
So one last comment here, and we can move it.

I understand the tables are different, but in the `split_into_pages_by_var` the listings *aren't* different, they're the same listing, but we're not guaranteed that `propose_column_widths` will give identical widths, so there you'd have a single listing, split by a variable, but the columns aren't guaranteed to line up

Reply:
My objections stand but I've removed the code from this PR as requested